syncserver/src/lib.rs

```diff
@@ -3,12 +3,243 @@
+                    $settings.statsd_port,
+                )
+                .expect("failed to build tokenserver configurator"),
+            )
+            .configure(configure_dockerflow)
+    };
+}
+
+#[cfg(all(feature = "syncstorage", feature = "tokenserver"))]
+#[macro_export]
+macro_rules! build_app {
+    ($settings: expr, $blocking_threadpool: expr) => {
+        App::new()
+            .data(Arc::new($settings.master_secret.clone()))
+            .service(
+                web::scope("/token")
+                    .wrap_fn(tokenserver_web::middleware::handle_request_log_line)
+                    .configure(
+                        tokenserver_web::get_configurator(
+                            &$settings.tokenserver,
+                            $settings.statsd_host.as_deref(),
+                            $settings.statsd_port,
+                            $blocking_threadpool,
+                        )
+                        .expect("failed to build tokenserver configurator"),
+                    ),
+            )
+            .service(
+                web::scope("/storage")
+                    .wrap(ErrorHandlers::new().handler(
+                        StatusCode::NOT_FOUND,
+                        syncstorage_web::middleware::render_404,
+                    ))
+                    // TODO: does this do what it's supposed to?
+                    .wrap(syncstorage_web::middleware::WeaveTimestamp::default())
+                    .configure(
+                        syncstorage_web::get_configurator(
+                            &$settings.syncstorage,
+                            $settings.statsd_host.as_deref(),
+                            $settings.statsd_port,
+                            $blocking_threadpool,
+                        )
+                        .expect("failed to build syncstorage configurator"),
+                    ),
+            )
+            // Middleware is applied LIFO
+            // These are our wrappers
+            // Followed by the "official middleware" so they run first.
+            // actix is getting increasingly tighter about CORS headers. Our server is
+            // not a huge risk but does deliver XHR JSON content.
+            // For now, let's be permissive and use NGINX (the wrapping server)
+            // for finer grained specification.
+            .wrap(build_cors($settings))
+    };
+}
+
+/// This function configures the following endpoints:
+/// * GET /__error__
+/// * GET /__version__
+/// * If syncstorage and tokenserver are both enabled, `GET __heartbeat__`, since we need to
+///   combine the `GET __heartbeat__` endpoints configured by syncstorage and tokenserver
+///   in their `get_configurator` functions
+///
+/// Any other dockerflow endpoints are configured individually by syncstorage and tokenserver.
+#[cfg(not(all(feature = "syncstorage", feature = "tokenserver")))]
+fn configure_dockerflow(cfg: &mut ServiceConfig) {
```

Comment:
The dockerflow endpoints that are general enough to be defined here are defined here